### PR TITLE
Add file dialog buttons to change book status

### LINKED ISSUE #8954: FR: add “Book status” or “Mark as…” to History and File Manager
**Does your feature request involve difficulty completing a task? Please describe.**
Now that we have a very useful sorting mechanism in History by books on hold, I think it would be handy to be able to change the status from History and File Manager. 

**Describe the solution you'd like**
History’s context menu has Book description and Book information. Let’s add Book status as well. File Manager has “Mark as finished”, but not other statuses. 

**Describe alternatives you've considered**
The only way to modify these statuses is by going to Book status, but only after opening the book. 

**Additional context**
A use case maybe when the user opens and skims a few books and the decides to put some of them on hold, then sort History by “Reading” and focus on a few more important books first. 

This is not a deal breaking omission, but I think such improvement will make my beloved KOReader even more useful tool for reading books. 

Thank you!

### PR DESCRIPTION
**Pull request by melyux**
Closes #8954.

Change is made to file manager, history, and collections alike. Most time-consuming part was understanding the inner workings of coverbrowser to allow compatibility with it.

Lots of code duplication due to historical reasons causes changes to look larger than they are, when it's mostly the same thing copied across the 3 view modules, and coverbrowser further trying to overwrite each separately. The special attention given to coverbrowser is simply because of the special cache it uses...

Ridealong tweaks:
- Remove the "Mark as reading/read" button from File Manager as it's superseded by this
- Add "Reset settings" button to Collections viewer to parallel File Manager and History, and to avoid having the ugly 2-button, 1-button, 2-button layout
- Rename "Sort" button in Collections viewer to be "Sort collection" to clarify that it's for sorting all files, not just the one being held down on
- Use existing `getButtonById` for fudging the "Book information" button callback instead of previously raw indexing into the buttons table, to be more resilient to changes that would bork the indexing (like these)
  - Fix this previously non-working fudging

### File manager buttons:
file manager buttons

### History buttons:
history buttons

### Favorites/Collections buttons:
favorites buttons

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9953)

### CODE REVIEW

File: frontend/apps/filemanager/filemanager.lua
Comment (melyux) [inline frontend/apps/filemanager/filemanager.lua:345]: We're already inside the `is_file` block

File: frontend/apps/filemanager/filemanagerhistory.lua
Comment (melyux) [inline frontend/apps/filemanager/filemanagerhistory.lua:92]: i.e. not deleted

File: frontend/apps/filemanager/filemanager.lua
Comment (melyux) [inline frontend/apps/filemanager/filemanager.lua:380]: We're already inside the `is_file` block

File: plugins/coverbrowser.koplugin/covermenu.lua
Comment (melyux) [inline plugins/coverbrowser.koplugin/covermenu.lua:441]: Moved this down after "Create the new ButtonDialog, and let UIManager show it" below, otherwise this newly overwritten callback is reset when that block runs

**Comment by mergen3107** [comment]
@melyux
Thank you!!

**Comment by hius07** [comment]
I thought of separate buttons to change the book status, but then decided to use our standard BookStatusWidget.

**Comment by melyux** [comment]
Competition @hius07? :P

I'll argue for treating setting the "readness" of books as a first class operation though, certainly on the level of "Cut" and "Book description".

Defer to the issue author? :) What say you @mergen3107

**Comment by mergen3107** [comment]
@melyux 
Yes, I very much like your current solution. As I mentioned in another thread somewhere, it would be nice to be able to skim through folders of new and old books, quickly give them statuses like finished/will read in future/interested now, and then start reading them from the History filters. Like putting tasks in baskets from all sources and then dealing with baskets. 

This PR does exactly this :)

**Comment by mergen3107** [comment]
From other reading-oriented perspectives, I liked one idea from a different app, that does this:
When you click a book in the library, it doesn’t open the book yet, but shows a window with book summary (title, author, cover, annotation, etc) and buttons on the bottom like “read now”, “manage favorites” and exit. 

Maybe we can do (optionally) something similar using the Book status? We would also have the 5-star rating (which I still think should be visible in the FM).

**Comment by NiLuJe** [comment]
> I'll argue for treating setting the "readness" of books as a first class operation though, on the level of "Cut" and "Book description". Fewer taps. More book-oriented, only takes up 1 more space

Oh, yeah, I would *definitely* hate to go through the BookStatusWidget thingy (which I despise with a fiery passion, but that' sjust me ;p), I missed this bit in #9956.

**Comment by melyux** [comment]
Yeah it's pretty low on my list of favorite widgets... don't mind it as a gateway into a book like imagined above tho

**Comment by poire-z** [comment]
Not sure how these new 3 items would look in other translations:
image
(I expect Cut/Copy/Rename to also be short in many languages, dunno about these 3 reading/hold/finished.)

I also don't use them, so they will feel like noise to me - but OK, I understand people needing quick access to them.

> I would definitely hate having to go through the BookStatusWidget thingy

May be it needs a non-fullscreen smaller quick access/use MiniBookStatusWidget ? :)
Or a single button cycling through the 3 states ?

**Comment by NiLuJe** [comment]
> Or a single button cycling through the 3 states ?

That is one of the reason why i never touched the "please add an explicit <whatever>" reading state: keeping with a binary read/unread status allowed us to get away with both simple code logic throughout, and a single stateful button.

Now, I understand the reasoning behind said requests, and, if push comes to shove, nnnyeeah, why not. But, yeah, I'd definitely try to think of something cleverer for the buttons in this case ;).

**Comment by melyux** [comment]
Hmmm maybe better to be explicit with the buttons, I can foresee it being annoying to have to go through the file dialog twice to get to the status you want. I liked that the 3 buttons fit well under the cut-delete-rename trio :)

**Comment by NiLuJe** [comment]
Those are valid points (and I am somewhat less concerned about the overall cramped feeling of the popup than @poire-z might be... as long as it still fits @ 800x600 167dpi ;)).

I went with "Mark as..." specifically to help with the read/read confusion, so, if we go the shorter route, I definitely agree we should find something better ;).

**Comment by mergen3107** [comment]
How about check marks?
Like, “ (box) Reading, (box) On Hold, (checked box) Finished”. All of them not grayed out, but the check mark always shows what’s the status.

**Comment by NiLuJe** [comment]
That would look fairly odd, doesn't help with the crowded feeling( might even make it worse), and I don't think we can actually do that without writing a brand new Button widget anyway ;o).

**Comment by poire-z** [comment]
> And also as a side effect, gives us at a quick glance the explicit status of the book (in case you have the dogear turned off): whatever's grayed out is the current status
> 
> Maybe "Reading", "On hold", "Read" as the labels to strengthen that effect and shorten the labels (notwithstanding the small problem of the confusion between "Read" the adjective and "Read" the verb, which could be solved by changing it to "Finished")

Could indeed look nice and short. `Reading` `On hold` `Finished`

(Not sure I would like it as it might be too standing out, but we could try having the current status with a gray (sic :) background ? image - mockup hacking the Font hinting toggle. Probably not worth trying as it may mess with the flashing button stuff.)

**Comment by melyux** [comment]
Is that still for the buttons? We can change the button backgrounds? Quite ambitious for a guy with flashing grays ;)

**Comment by poire-z** [comment]
> We can change the button backgrounds? 

May be not in nature, where buttons, like sheeps, are white.
But with the power of the human mind and heavy use of technology, a little dream can become real !
image

> Quite ambitious for a guy with flashing grays ;)

I was indeed just speaking in the abstract. I would advise to not go against nature and its black and white :)

**Comment by NiLuJe** [comment]
What do we need to do to get the ball rolling again on this (and/or #9956, @hius07)?

**Comment by melyux** [comment]
I was going to change the button names to "Reading", "On hold", "Finished" to address the length concerns for other languages. I'll do it and post a picture later today

Otherwise, some spitballing about gray backgrounds vs the currently default behavior of grayed out text when a file dialog button is already active and can't be selected

**Comment by hius07** [comment]
The code for 3 buttons is similar, can be optimized to `local function genStatusButton(status)` in three modules.

**Comment by melyux** [comment]
Shortened the names
FileManager_2023-02-03_235128 2

@hius07 because we're not touching the old "reading", "abandoned", "complete" status aliases, we'd end up with a lot of if and else statements to get the right text, id, status for each status that you can't predict from the input `status` arg. It may not be worth it. We could add a function in `filemanagerutil` to generate the button group once so we can call it from `filemanager` and `filemanagercollection` and `filemanagerhistory` if we really want, that might work better

**Comment by hius07** [comment]
You can make `id = status` and have a `status_text` table of 3 keys like `{abandoned = _("On hold")}`.

**Comment by melyux** [comment]
Having some trouble with implementing it, since the button callback relies on the lexical scoping for a lot of things. I could define the callback in the calling function, but that'd be messy and gets me diminishing returns. Not sure what to do, can I get a coding assist? @hius07

**Comment by hius07** [comment]
The idea is something like
```Lua
local function genStatusButton(status)
    local status_text = {
        abandoned = _("On hold"),
        reading   = _("Reading"),
        complete  = _("Finished"),
    }
    return {
        text = status_text[status],
        id = status,
        enabled = status ~= current_status,
        callback = function()
            filemanagerutil.setStatus(file, status)
            self:refreshPath()
            UIManager:close(self.file_dialog)
        end,
    }
end

```

File: frontend/apps/filemanager/filemanagerhistory.lua
Comment (hius07) [reply frontend/apps/filemanager/filemanagerhistory.lua:92]: I think you have item.dim.

**Comment by hius07** [comment]
Are you going to set the grey background for a button with current status (disabled)? If so, it would be better to not show status buttons for deleted files in History at all. Otherwise all three buttons would have a grey bg.

**Comment by NiLuJe** [comment]
I thought we'd concluded that the gray background was a terrible idea? ^^ (hint: it is ;p).

**Comment by hius07** [comment]
> the button callback relies on the lexical scoping for a lot of things

You are right, changed the example a little bit.

File: frontend/apps/filemanager/filemanagerhistory.lua
Comment (melyux) [reply frontend/apps/filemanager/filemanagerhistory.lua]: Nice, done. The name threw me

**Comment by melyux** [comment]
The original example works though, how's this?

**Comment by hius07** [comment]
As you said, we can go further and move genStatusButton to filemanagerutil. The changed example is for it.

**Comment by melyux** [comment]
Had to move the enabled logic to the arguments b/c of history's additional item.dim check. How's this?

**Comment by hius07** [comment]
Some ideas on further optimization
- `getStatusButtonsRow` (return the full row of three buttons)
- for deleted files in History do not show the row of status buttons (all disabled) at all
- pass `current_status` (not the condition for `enabled`) as argument

I've missed, how the current status is indicated in the button (since the gray bg is bad)?

**Comment by melyux** [comment]
Done.

The current status right now is secondarily shown by which status button is grayed out

**Comment by hius07** [comment]
Some further minor optimizations
(1) In filemanagerutils please change in (my) line 62 `close` to `flush`.
(2) `current_status` can be get within `getStatusButtonsRow` and not passed there as argument.
(3) `genStatusButton` is better to be a local function within `getStatusButtonsRow`.
(4) How do you feel the more logical button placement: "On hold" the leftmost, "Reading" in the centre, "Finished" the rightmost? In any case, the placement must be the same in the Status widget and in the popup buttontable.
(5) "Reset settings" button is the same in FM, History and Collections. Maybe move it to the utils too? It seemes that the `caller_callback` for it is the same as for the status buttons?
(6) Please check the buttontable in History in classic mode. I mean the separator after the status buttons.
(7) covermenu: can the repeated code be done in the loop over `ipairs({"reading", "abandoned", "complete"}`?

**Comment by melyux** [comment]
1) Done
2) Done
3) Done
4) I think the current placement is more intuitive, and here's me trying to put it into words: Usually books are almost always "Reading" before they come "On hold", so it would be strange to have "On hold" be the first option, since the buttons are more or less in left to right reading progression order. You almost never start with a book "On hold" before "Reading"
5) Done. I also renamed "Reset settings" to "Reset", since this button does more than just reset settings—it resets everything about the book including reading progress.
5b) Found that Collections/Favorites was missing the check for History that prevents you from Resetting a currently open file. Added that check in
6) What about it? It looks as expected. What'd you see?
7) Going to try this in a separate commit

**Comment by hius07** [comment]
> 6\. What about it? It looks as expected. What'd you see?

It's okay, nevermind.

File: frontend/apps/filemanager/filemanagerutil.lua
Comment (hius07) [inline frontend/apps/filemanager/filemanagerutil.lua:116]: leftover

File: frontend/apps/filemanager/filemanagerutil.lua
Comment (hius07) [inline frontend/apps/filemanager/filemanagerutil.lua:135]: Since its local now, you can pass `to_status` only.

File: frontend/apps/filemanager/filemanagerutil.lua
Comment (melyux) [reply frontend/apps/filemanager/filemanagerutil.lua:116]: Doh

File: frontend/apps/filemanager/filemanagerutil.lua
Comment (melyux) [reply frontend/apps/filemanager/filemanagerutil.lua:135]: V good

**Comment by melyux** [comment]
Besides the `ipairs({"reading", "abandoned", "complete"}` optimization, I see that that whole code block in covermenu for the files, history, and collections is almost identical and very long. Moving it into a method that they will all call

**Comment by melyux** [comment]
Done with the `ipairs` loop. Also fixed my misplacement in collections that was causing a crash, and took out all args except `to_status` from the inner `genStatusButton` function since it doesn't need them anymore

File: frontend/apps/filemanager/filemanagerutil.lua
Comment (hius07) [inline frontend/apps/filemanager/filemanagerutil.lua:116]: Maybe `gen`? (generate)

Comment (melyux) [reply frontend/apps/filemanager/filemanagerutil.lua:116]: Hmm?

**Comment by hius07** [comment]
```Lua
function CoverMenu:updateCache(file, status)
    if self.cover_info_cache and self.cover_info_cache[file] then
        if status then
            self.cover_info_cache[file][3] = status
        else
            self.cover_info_cache[file] = nil
        end
    end
end
```

**Comment by melyux** [comment]
@hius07 ??? You mean use this 

> ```lua
> function CoverMenu:updateCache(file, status)
>     if self.cover_info_cache and self.cover_info_cache[file] then
>         if status then
>             self.cover_info_cache[file][3] = status
>         else
>             self.cover_info_cache[file] = nil
>         end
>     end
> end
> ```

I tried replacing all those usages with a call to this method, but getting `attempt to call method 'updateCache' (a nil value)` when it's called. How would you call this method? `self:updateCache(file, status)` doesn't work for because the `self` is the original menu, not CoverBrowser

**Comment by NiLuJe** [comment]
`CoverMenu.updateCache(self, file, status)` ought to work, as would something along the lines of `self.updateCache = CoverMenu.updateCache` somewhere in the affected modules?

**Comment by melyux** [comment]
Does the method above have to be changed for `CoverMenu.updateCache(self, file, status)` to work? It's not erroring out anymore but it's not actually doing anything either @NiLuJe

**Comment by NiLuJe** [comment]
How does the call look, in context?

(Sidebar: the second approach of just pointing the class methods to CoverMenu's would *probably* be less confusing ;)).

**Comment by NiLuJe** [comment]
Oh, wait, this needs more trickery, the object that holds the `cover_info_cache` field is always CoverMenu's instance, not the caller's.

**Comment by hius07** [comment]
Look at the main.lua of this plugin, to see how the methods of CoverMenu are inserted into self.

**Comment by NiLuJe** [comment]
> Oh, wait, this needs more trickery, the object that holds the `cover_info_cache` field is always CoverMenu's instance, not the caller's, right?

At a (very, very) quick glance, currently, the callers are `ListMenuItem` or `MosaicMenuItem` instances, and they poke at their parent's object via `self.menu` to access `self.menu.cover_info_cache`. `CoverMenu` is a `Menu`, not a `MenuItem`, so there's no extra pointer lookup, it's `self.cover_info_cache`.

You either need two different methods, or a hybrid one that does something along the lines of

```
local cover_info_cache = self.menu and self.menu.cover_info_cache or self.cover_info_cache
```

first, and operates on that pointer instead.

**Comment by melyux** [comment]
Simply inserting the new method like this into the 3 locations in `main.lua` _seems_ to work, per @hius07. 

@NiLuJe Are the callers really `ListMenuItem` etc? I really can't make out all the crazy logic in the code :( It does seem to be working without the hybrid call though...

File: frontend/apps/filemanager/filemanagerutil.lua
Comment (hius07) [reply frontend/apps/filemanager/filemanagerutil.lua:116]: I mean `genStatusButtonsRow`.

Comment (melyux) [reply frontend/apps/filemanager/filemanagerutil.lua]: Good eye, didn't even see it until looking at it 10 times

**Comment by NiLuJe** [comment]
> @NiLuJe Are the callers really `ListMenuItem` etc?

I'm basing this on the state of the code in master and a quick grep, you may be doing something slightly different than I thought ;o).

**Comment by NiLuJe** [comment]
In any case, that commit looks good ;).

**Review by poire-z — APPROVED** [review]
Just had a quick glance at the code, but nothing jumps out.
Tested on my Kobo, and it looks ok and seems to work fine.

**Comment by Frenzie** [comment]
@hius07 Good for you?

**Comment by hius07** [comment]
Yes I think it's good.

**Review by NiLuJe — COMMENTED** [review]
Late to the party, sorry; had to step away earlier ;).

> Reviewed 1 of 5 files at r1, 3 of 6 files at r3, 2 of 2 files at r10, 3 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9953#-:-NO1c8a-262lmhR2y5ZW:bob8xvd)* status: 8 of 9 files reviewed, 7 unresolved discussions

___
*[`frontend/apps/filemanager/filemanagerutil.lua` line 79 at r11](https://reviewable.io/reviews/koreader/koreader/9953#-NO1_WZrAv0LUBp_Sv9H:-NO1_WZrAv0LUBp_Sv9I:bu0zs68) ([raw file](https://github.com/koreader/koreader/blob/d6f5294ded19f4dd117fc7040de26da56a62197b/frontend/apps/filemanager/filemanagerutil.lua#L79)):*
> ```Lua
>         else
>             status = "reading"
>         end
> ```

In the spirit of #10105 

```lua
local doc_settings = DocSettings:open(file)
local summary = doc_settings:readSetting("summary")
if summary and summary.status ~= "" then
    status = summary.status
else
     status = "reading"
end
```

(There may be a few extra and useless `summary.status` checks before making a string comparison in other parts of the code. This is unnecessary. t.foo yields `nil`, so it's not necessary to explicitly check for it first, you can compare `nil` with a string).



_Code quote:_
```Lua
        local docinfo = DocSettings:open(file) -- no io handles created, do not close
        if docinfo.data.summary and docinfo.data.summary.status and docinfo.data.summary.status ~= "" then
            status = docinfo.data.summary.status
        else
            status = "reading"
        end
```
___
*[`frontend/apps/filemanager/filemanagerutil.lua` line 110 at r11](https://reviewable.io/reviews/koreader/koreader/9953#-NO1b0Rs6rX9wxqW9jpo:-NO1b0Rs6rX9wxqW9jpp:b-lve42x) ([raw file](https://github.com/koreader/koreader/blob/d6f5294ded19f4dd117fc7040de26da56a62197b/frontend/apps/filemanager/filemanagerutil.lua#L110)):*
> ```Lua
>             }
>         end
>     end
> ```

Ditto on the `DocSettings` API usage instead of poking at the data filed manually here ;).



_Code quote:_
```Lua
    -- In case the book doesn't have a sidecar file, this'll create it
    local docinfo = DocSettings:open(file)
    local summary
    if docinfo.data.summary and docinfo.data.summary.status then
        -- Book already had the full BookStatus table in its sidecar, easy peasy!
        docinfo.data.summary.status = status
        docinfo.data.summary.modified = os.date("%Y-%m-%d", os.time())
        summary = docinfo.data.summary
    else
        -- No BookStatus table, create a minimal one...
        if docinfo.data.summary then
            -- Err, a summary table with no status entry? Should never happen...
            summary = { status = status }
            -- Append the status entry to the existing summary...
            require("util").tableMerge(docinfo.data.summary, summary)
            docinfo.data.summary.modified = os.date("%Y-%m-%d", os.time())
            summary = docinfo.data.summary
        else
            -- No summary table at all, create a minimal one
            summary = {
                status = status,
                modified = os.date("%Y-%m-%d", os.time())
            }
        end
    end
```

**Comment by hius07** [comment]
> In the spirit of #10105

I've seen it but left to myself.

**Comment by hius07** [comment]
Besides, https://github.com/koreader/koreader/issues/9952 must be fixed.
(I've closed https://github.com/koreader/koreader/pull/10011 in favor of a new complex PR).

**Comment by poire-z** [comment]
(@Frenzie : beware your squash/merge button :), that's 17 commits into master, that although they are verbose english, are a bit overwhelming without context.)

**Comment by Frenzie** [comment]
It didn't seem right to squash them, but yes, the commit messages are not completely ideal.

**Comment by mergen3107** [comment]
@melyux 
Buttons work absolutely fantastically! Thank you!
I checked in History and in FM. 

Next evolution step might be add these buttons to the multi-file selection experience (just a thought).

**Comment by hius07** [comment]
> Next evolution step might be add these buttons to the multi-file selection experience (just a thought).

Please take a little break pending new Docsettings https://github.com/koreader/koreader/pull/10074 and its support everywhere is done.

**Comment by mergen3107** [comment]
@hius07 
Thank you! I can’t wait to test your new PR as well!

**Comment by mergen3107** [comment]
@melyux 
One thought: without the context, it is difficult to understand what these three buttons mean from a glance. 

I can see that the three status buttons are separated in a group, but I don’t think this is obvious enough (at least, for newcomers). How about adding a single thick border around the group?

Or, better, is there a possibility to add a text string saying “Book status” or something like that on on of them?

**Comment by poire-z** [comment]
> but I don’t think this is obvious enough (at least, for newcomers). How about adding a single thick border around the group?
> Or, better, is there a possibility to add a text string saying “Book status” or something like that on top of the group?

Don't understimate newcomers :)
(And we would be tempted then to add "File operations", "Metadata" around the other groups. :)
I'd rather keep this popup small and lean.

**Comment by mergen3107** [comment]
Ok, then maybe change the word “Reading” to “Being read”?
Let me explain my stand of view. 

So your open the long tap and you can:
Copy (the book) - OK
Delete (the book) - Ok
etc

or, 
Reset settings (of the book) - also OK. 
etc

Now we come to the status:
(The book is) Finished - OK
(The book is) On hold - OK
(The book is) Reading? Book is not reading per se, book is “being read”.
If you make it “(I am) Reading (the book)” it makes sense, but it is now out of context with respect to the other two. 

Another alternative is to return the verbs:
Mark as reading
Put on hold
Mark as finished

Again, personally I don’t have anything against current implementation, it works great! Just wanted to share my thought while the coal is hot so we can discuss now :D

**Comment by melyux** [comment]
The verbs were clearer, true. Concern was that their translation in another language may not have fit into the button box

**Comment by mergen3107** [comment]
Thanks! In that context, how about “Reading” -> “In progress”?

**Comment by poire-z** [comment]
Not using the status, so I want them small and not in my face :) so, happy with the current wording.
I've seen this discussed above (or elsewhere), but if the popup status button ordering has been thought of, the BookStatusWidget (which is, or will be, less often used) should now get them in the same order - so it's consistent (and may help newcomers to figure "hey, I've seen them in the nice bookstatuswidget, these are shortcuts!).

> Thought you guys would squash the commits like we did before, otherwise I would have been more hygienic about them 

It's good to keep all your 17 commits in the PR unsquashed, and it's indeed our job to squash them. It is sometimes a bit hard, like with this PR, when there are many commits, some of them having verbose descriptions (that could end being kept in the squash commit message), and when the feature has been discussed and modified all along the PR (and the individual commit messages are no longer true): it's _really_ hard for us to write the final commit message stating the end result, even more when we were not really involved in the discussion :)
So, I may sometime request the PR author to write that message for me :)

**Comment by Frenzie** [comment]
@melyux Looking at some older PRs I see your style always communicates to me that you intend your commits to be rebased. I made the decision not to nitpick.

**Comment by mergen3107** [comment]
I got another thought about status button text. How about putting square brackets or quotation marks?
Like,
`[Reading]`, `[On hold]`, `[Finished]`
or
`”Reading”`, `”On hold”`, `”Finished”` 

So that these buttons suggest that these word are not actions (as Copy, for example), but something else, like statuses

**Comment by NiLuJe** [comment]
I feel like the fact that they're already on their own island surrounded by horizontal separators is enough?

**Comment by mergen3107** [comment]
> I can see that the three status buttons are separated in a group, but I don’t think this is obvious enough

Doesn’t look like to me :D

**Comment by melyux** [comment]
But they _are_ actions, they mark the book as that status

**Comment by mergen3107** [comment]
Yes, they are. My point is that the words do not reflect that well. They are “Mark as …” actions

**Comment by hius07** [comment]
It can be a small checkmark indicating the current status (disabled button) and helping to visually distinguish the row:

**Comment by melyux** [comment]
Checkmark looks great

**Pull request merged**
